src/main/java/org/sonarsource/scanner/maven/bootstrap/MavenCompilerResolver.java

```diff
@@ -0,0 +1,298 @@
+    Xpp3Dom configuration = exec.getConfiguration();
+    PlexusConfiguration pomConfiguration = new XmlPlexusConfiguration(configuration);
+    PlexusConfiguration config = pomConfiguration.getChild(parameterName, false);
+    if (config == null) {
+      return Optional.empty();
+    }
+    return Optional.ofNullable(convertString(exec, config));
+  }
+
+  private String convertString(MojoExecution exec, PlexusConfiguration config) {
+    ExpressionEvaluator expressionEvaluator = new PluginParameterExpressionEvaluator(session, exec);
+    BasicStringConverter converter = new BasicStringConverter();
+    try {
+      return converter.fromExpression(config, expressionEvaluator);
+    } catch (ComponentConfigurationException e) {
+      log.warn(e.getMessage(), e);
+      return null;
+    }
+  }
+
+  private Optional<Map<String, String>> getMapConfiguration(MojoExecution exec, String parameterName) {
+    Xpp3Dom configuration = exec.getConfiguration();
+    PlexusConfiguration pomConfiguration = new XmlPlexusConfiguration(configuration);
+    PlexusConfiguration config = pomConfiguration.getChild(parameterName, false);
+    if (config == null) {
+      return Optional.empty();
+    }
+    return Optional.of(Stream.of(config.getChildren())
+      .collect(Collectors.toMap(PlexusConfiguration::getName, c -> convertString(exec, config.getChild(c.getName(), false)))));
+  }
+
+  private Optional<String> getJdkHome(MojoExecution compilerExecution) {
+
+    Optional<String> executable = getStringConfiguration(compilerExecution, "executable");
+    if (executable.isPresent()) {
+      // Should be something like <jdk home>/bin/javac
+      return Optional.of(Paths.get(executable.get()).getParent().getParent().toString());
+    }
+
+    // Inspired by
+    // https://github.com/apache/maven-compiler-plugin/blob/dc4a5635ba4eb2ba5e461fa53b2c47c58d7fa397/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L1418
+    Toolchain tc = null;
+
+    // Maven 3.3.1 has plugin execution scoped Toolchain Support
+    Optional<Map<String, String>> jdkToolchain = getMapConfiguration(compilerExecution, "jdkToolchain");
+    if (jdkToolchain.isPresent() && !jdkToolchain.get().isEmpty()) {
+      List<Toolchain> tcs = collectMatchingToolchains(jdkToolchain.get());
+      if (!tcs.isEmpty()) {
+        tc = tcs.get(0);
+      }
+    }
+
+    // Fallback on the global jdk toolchain
+    if (tc == null) {
+      tc = toolchainManager.getToolchainFromBuildContext("jdk", session);
+    }
+
+    if (tc instanceof DefaultJavaToolChain) {
+      return Optional.of(((DefaultJavaToolChain) tc).getJavaHome());
+    }
+
+    // Like m-compiler-p, last fallback is to compile with the runtime JDK
+    String runtimeCompiler = getJavacExecutableFromRuntimeJdk();
+    if (runtimeCompiler != null) {
+      return Optional.of(Paths.get(runtimeCompiler).getParent().getParent().toString());
```

Comment:
This logic could be extracted into a dedicated method: `Optional<String> getJdkHomeFromJavac(String compiler) {...}` and used here and on line 209

Reply:
:+1: